crypto/ec/curve448/f_generic.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2015-2016 Cryptography Research, Inc.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ * Originally written by Mike Hamburg
+ */
+#include "field.h"
+
+static const gf MODULUS = {
+    FIELD_LITERAL(0xffffffffffffff, 0xffffffffffffff, 0xffffffffffffff,
+                  0xffffffffffffff, 0xfffffffffffffe, 0xffffffffffffff,
+                  0xffffffffffffff, 0xffffffffffffff)
+};
+
+/* Serialize to wire format. */
+void gf_serialize(uint8_t serial[SER_BYTES], const gf x, int with_hibit)
+{
+    unsigned int j = 0, fill = 0;
+    dword_t buffer = 0;
+    int i;
+    gf red;
+
+    gf_copy(red, x);
+    gf_strong_reduce(red);
+    if (!with_hibit)
+        assert(gf_hibit(red) == 0);
+
+    UNROLL for (i = 0; i < (with_hibit ? X_SER_BYTES : SER_BYTES); i++) {
+        if (fill < 8 && j < NLIMBS) {
+            buffer |= ((dword_t) red->limb[LIMBPERM(j)]) << fill;
+            fill += LIMB_PLACE_VALUE(LIMBPERM(j));
+            j++;
+        }
+        serial[i] = (uint8_t)buffer;
+        fill -= 8;
+        buffer >>= 8;
+    }
+}
+
+/* Return high bit of x = low bit of 2x mod p */
+mask_t gf_hibit(const gf x)
+{
+    gf y;
+    gf_add(y, x, x);
+    gf_strong_reduce(y);
+    return 0 - (y->limb[0] & 1);
+}
+
+/* Return high bit of x = low bit of 2x mod p */
+mask_t gf_lobit(const gf x)
+{
+    gf y;
+    gf_copy(y, x);
+    gf_strong_reduce(y);
+    return 0 - (y->limb[0] & 1);
+}
+
+/* Deserialize from wire format; return -1 on success and 0 on failure. */
+mask_t gf_deserialize(gf x, const uint8_t serial[SER_BYTES], int with_hibit,
+                      uint8_t hi_nmask)
+{
+    unsigned int j = 0, fill = 0;
+    dword_t buffer = 0;
+    dsword_t scarry = 0;
+    const unsigned nbytes = with_hibit ? X_SER_BYTES : SER_BYTES;
+    unsigned int i;
+    mask_t succ;
+
+    UNROLL for (i = 0; i < NLIMBS; i++) {
```

Comment:
yet another differnet way of loop unrolling.
this one is limited to clang, so I have to object :)

Reply:
how about this:

```
#define FOR_LIMB(_i, _start, _end, _x) \
    do { \
      (_i) = (_start); \
      do { \ 
        switch((_end) - (_i)) { \
        default: \
        case 8: _x; (_i)++; /* fall through */ \
        case 7: _x; (_i)++; /* fall through */ \
        case 6: _x; (_i)++; /* fall through */ \
        case 5: _x; (_i)++; /* fall through */ \
        case 4: _x; (_i)++; /* fall through */ \
        case 3: _x; (_i)++; /* fall through */ \
        case 2: _x; (_i)++; /* fall through */ \
        case 1: _x; (_i)++; continue; \
        case 0: break; \
        } \
     } while(0); \
    } while(0)
```
could translate
`UNROLL while (fill < LIMB_PLACE_VALUE(LIMBPERM(i)) && j < nbytes) {`
to
`FOR_LIMB(j, j, nbytes, { if (fill < LIMB_PLACE_VALUE(LIMBPERM(i))) break; ...`

Reply:
Aehm, correction:

```
        case 1: _x; (_i)++; continue; \
        case 0: break; \
        } \
     break; \
     } while(1); \
   } while(0)
```

Reply:
Done